### PR TITLE
project-survey-jenny-laura

### PR DESCRIPTION
**Pull request by LauraLyckholm**
See live here:

https://morning-person-quiz.netlify.app/

### CODE REVIEW

File: src/app.css [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    display: flex;
+    flex-direction: column;
+}
+
+.progress-bar {
+    font-style: italic;
+    opacity: 0.5;
+}
+
+/* === Form elements === */
```
Comment (IdahCollin) [inline src/app.css:67]: "Sectioning" the comments in this CSS makes it easier to follow the code when it's all in the same CSS. Good job!

File: src/components/Answers/AnswerTemplates.jsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
+// Template with answers depending on the chosen answers
+export const textTemplates = {
+    // 1-exercising-salty
+    '1-exercising-salty': `Hello [Name]! You're a caffeine enthusiast morning person who loves to kick off the day with exercise. For breakfast, you enjoy a salty combo. You've got the perfect recipe for a slow and flavorful morning.`,
```
Comment (IdahCollin) [inline src/components/Answers/AnswerTemplates.jsx:4]: Fun way with all the different combinations to make it feel personalized!

File: src/components/Questions/ProgressBar.jsx [hunk 3]
```diff
@@ -0,0 +1,5 @@
+export const ProgressBar = ({ currentStep, questionsLength }) => {
```
Comment (IdahCollin) [inline src/components/Questions/ProgressBar.jsx:1]: Great job adding the progress bar!

File: src/components/Questions/QuestionComponent.jsx [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    setUserAnswers((prevAnswers) => { // To make sure we are working with the newest data
+      const updatedAnswers = [...prevAnswers]; // Creates a copy of the previous answers array (prevAnswers) to avoid mutating the state directly
+      updatedAnswers[currentStep] = updatedAnswer; // Updates answers array with the new answer, the one relevant for the current step
+      return updatedAnswers;
+    });
+  };
+
+  // Function to handle moving to the next question
+  const handleNextQuestionClick = () => {
+    // Validate user input for the name question
```
Comment (IdahCollin) [inline src/components/Questions/QuestionComponent.jsx:26]: Great!

File: src/questions.json [hunk 5]
```diff
@@ -0,0 +1,65 @@
+[
```
Comment (IdahCollin) [inline src/questions.json:1]: Give you an applause for this extra .json-file!

**Review by IdahCollin — APPROVED** [review]
Great job Jenny and Laura! The code works, fulfills requirements (and then some!), follows guidelines and is described. I really enjoyed your work!

As you wrote in the 'read-me' more components sounds like a nice next step. Great that you commented the css since it's in the same file. Nice that you worked out the differences in your schedules and used it as an advantage for you.

//Idah

File: src/App.jsx [hunk 6]
```diff
@@ -1,3 +1,79 @@
...
+import { QuestionComponent } from './components/Questions/QuestionComponent'; // Logic for displaying questions
+import { AnswersSummary } from './components//Answers/AnswersSummary'; // AnswersSummary component
+import { ProgressBar } from './components/Questions/ProgressBar';
+
+// Main component that renders all the Quiz Components
 export const App = () => {
-  return <div>Find me in src/app.jsx!</div>;
-};
+  // State to track the current question step, user answers, and completion status
+  const [currentStep, setCurrentStep] = useState(0);
```
Comment (emmydieden) [inline src/App.jsx:12]: I like how you have set all the states in the parent component and then passed them on to the children components. Makes it easier to follow.

File: src/App.jsx [hunk 7]
```diff
@@ -1,3 +1,79 @@
...
+          />
+          <ProgressBar currentStep={currentStep} questionsLength={questions.length} />
+        </section>
+      ) : (
+        // Display the loading or answer summary section after completing the form
+        <section id="answers-summary">
+          {showLoading ? (
+            // Display a loading message or icon here
+            <div className="loading">
+              <p>Your answer is being prepared...</p>
```
Comment (emmydieden) [inline src/App.jsx:67]: Love this little detail (the loading message and icon) in your survey!

File: src/components/Answers/AnswerTemplates.jsx [hunk 8]
```diff
@@ -0,0 +1,113 @@
+// Template with answers depending on the chosen answers
+export const textTemplates = {
```
Comment (emmydieden) [inline src/components/Answers/AnswerTemplates.jsx:2]: Clever to make templates for the questions and put them in their own component. Makes it easier to handle the end result and keeps rest of the code cleaner. And makes it possible for you to have so many custom made answers! Very nice addition to the survey.

File: src/components/Answers/AnswersSummary.jsx [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { textTemplates } from "./AnswerTemplates.jsx";
+import { Button } from "../FormElements/Button.jsx";
+
+export const AnswersSummary = ({ userAnswers, handleRestart }) => {
+  // Function to generate an answer based on user responses
+  const generateAnswer = () => {
+    // Defines the template keys based on user responses, name (index=[0]) is excluded here, since it's not present in the answers-template
+    const templateKey = `${userAnswers[1]}-${userAnswers[2]}-${userAnswers[3]}`;
```
Comment (emmydieden) [inline src/components/Answers/AnswersSummary.jsx:8]: This is great logic to get hold of the templateKey and match it to the right template!

File: src/components/Questions/ProgressBar.jsx (hunk 3 above)
Comment (emmydieden) [reply src/components/Questions/ProgressBar.jsx:1]: I agree with Idah here! And you've managed to keep the code pretty simple by importing the questions as an array and passing on the length of the array as a prop. Clever!

File: src/components/Questions/QuestionComponent.jsx [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    if (currentStep === 0) {
+      const userName = userAnswers[0];
+      if (!/^[a-zA-Z\söäåÖÄÅ]+$/.test(userName)) { // A regex to allow lowecase and uppercase letters, and special characters å ä ö, tests them against the text input into the name field. The exclamationmark is there to negate the test, meaning to allow for a valid input
```
Comment (emmydieden) [inline src/components/Questions/QuestionComponent.jsx:29]: Well done for adding a regex!

File: src/components/Questions/QuestionComponent.jsx [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    setValidationError('');
+
+    // Move to the next question
+    handleNextQuestion();
+  };
+
+  return (
+    <div className="question-section">
+      {/* Conditional rendering based on the current question and completion status */}
+      {currentStep < questions.length && !completed ? (
```
Comment (emmydieden) [inline src/components/Questions/QuestionComponent.jsx:51]: Great usage of ternary operators in the following code, although I find it quite hard to follow the logic - when different blocks of code are executed or not. I think this is the downside of ternary operators - they don't make it very easy to read the code as it's easy to miss the "?" and ":". But that's of course not your fault!:) But maybe a few more comments would have helped?

**Review by emmydieden — COMMENTED** [review]
Loved your survey! The design, the questions, the answers, the details - everything!:) 
You have really made the most out of this week's task to use state to make your app react to to the user interaction in different ways. You used states in a way I hadn't thought of - it was for example a great idea to set your answers as useState. 
You have used great ways to "make things easier" - you have the ready made answer templates and you imported the questions as a JSON. Things that keeps the rest of your code very clean. 
It was very easy to follow your code, due to good variable names and many explaining comments. The only time I got a bit lost was in the Question Component as very much logic in this component is executed by conditional ternary operators and I often find it hard to read the logic when it comes to these, as they are not as visually explanatory as for example if/else statements. But as I said in the comments, this is nothing you are responsible for;)

Thanks for letting me review your code!
